src/mlpack/methods/decision_tree/best_binary_numeric_split_impl.hpp

```diff
@@ -185,14 +182,103 @@ double BestBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
   return bestFoundGain;
 }
 
+template<typename FitnessFunction>
+template<bool UseWeights, typename VecType, typename WeightVecType>
+double BestBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
+    const double bestGain,
+    const VecType& data,
+    const arma::Row<double>& labels,
+    const size_t numClasses,
+    const WeightVecType& weights,
+    const size_t minimumLeafSize,
+    const double minimumGainSplit,
+    double& splitInfo,
+    AuxiliarySplitInfo& /* aux */)
+{
+  // First sanity check: if we don't have enough points, we can't split.
+  if (data.n_elem < (minimumLeafSize * 2))
+    return DBL_MAX;
+  if (bestGain == 0.0)
+    return DBL_MAX; // It can't be outperformed.
+
+  // Next, sort the data.
+  arma::uvec sortedIndices = arma::sort_index(data);
+  arma::Row<double> sortedLabels(labels.n_elem);
+  arma::rowvec sortedWeights;
+  for (size_t i = 0; i < sortedLabels.n_elem; ++i)
+    sortedLabels[i] = labels[sortedIndices[i]];
+
+  // Sanity check: if the first element is the same as the last, we can't split
+  // in this dimension.
+  if (data[sortedIndices[0]] == data[sortedIndices[sortedIndices.n_elem - 1]])
+    return DBL_MAX;
+
+  // Only initialize if we are using weights.
+  if (UseWeights)
+  {
+    sortedWeights.set_size(sortedLabels.n_elem);
+    // The weights must keep the same order as the labels.
+    for (size_t i = 0; i < sortedLabels.n_elem; ++i)
+      sortedWeights[i] = weights[sortedIndices[i]];
+  }
+
+  double bestFoundGain = std::min(bestGain + minimumGainSplit, 0.0);
+  bool improved = false;
+  // Force a minimum leaf size of 1 (empty children don't make sense).
+  const size_t minimum = std::max(minimumLeafSize, (size_t) 1);
+
+  // Loop through all possible split points, choosing the best one.
+  for (size_t index = minimum; index < data.n_elem - minimum + 1; ++index)
+  {
+    // Make sure that the value has changed.
+    if (data[sortedIndices[index]] == data[sortedIndices[index - 1]])
+      continue;
+
+    // Calculate the gain for the left and right child.
+    const double leftGain = FitnessFunction::template Evaluate<UseWeights>(sortedLabels,
+        sortedWeights, 0, index);
+    const double rightGain = FitnessFunction::template Evaluate<UseWeights>(sortedLabels,
+        sortedWeights, index, labels.n_elem);
```

Comment:
The core of the loop for `SplitIfBetter` is the same, but the different part is here, in how we compute the gain of the left and right child.  I might suggest instead splitting out the gain computation part, so that we have only one `SplitIfBetter`, but then when we compute the child gains, that is where we split out to different code.

We might consider, e.g., modifying the `FitnessFunction::Evaluate()` definition to something a bit more like this:

```
template<typename LabelsType, typename WeightsType, typename UseWeights>
FitnessFunction::EvaluateSplit(
    const LabelsType& labels,
    const WeightsType& weights,
    const size_t splitIndex,
    double& totalGain,
    double& leftGain,
    double& rightGain);
```

Let me know if that makes sense, and what you think of the idea.  I think it would be best if we managed to keep `BestBinaryNumericSplit` entirely independent of whether the task is regression or classification---we just push that detail down into the `FitnessFunction`.

Reply:
Apart from the difference in gain computation inside the for loop, the `SplitIfBetter` also keeps track of `classCounts` for each iteration.

So, the main idea for making two overloads here was that since finding the best numeric split is computationally expensive, so in future, we can store some information about labels like their prefix sum in order to speed up the process (like we are doing by storing `classCounts` in case of classification).

I did come up with an algorithm that could do optimization, but it was only applicable for training without weights. The mathematical equation for this optimization in weighted training was very complicated and I was unable to solve it. So, I decided that we can leave it for future if some better mathematician than me could figure out a way to do this for weighted training also :).

What you are suggesting might become a bit difficult to handle because we need to calculate `classCounts` only for classification and that too while calculating `BestNumericSplit`, so even though we change the `FitnessFunction::Evaluate`, then also that can't be used anywhere else apart from this.

I apologise if I overlooked something. Does my reasoning for making this choice makes sense?

Reply:
I see, I understand now.  Thanks for the explanation---I had missed `classCounts`.  I think the best option here is to have two separate functions, like have done it, then. :+1:

Reply:
Do you want to leave a TODO in the code comments about the optimization you were discussing?

Reply:
> I see, I understand now. Thanks for the explanation---I had missed classCounts. I think the best option here is to have two separate functions, like have done it, then. 👍

Awesome! :D

> Do you want to leave a TODO in the code comments about the optimization you were discussing?

Sure, I will do that. 👍